src/content_scripts/autocorrect.js

```diff
@@ -0,0 +1,351 @@
+
+const fractionValues = Object.freeze([1.0 / 4.0, 1.0 / 2.0, 3.0 / 4.0, 1.0 / 7.0, 1.0 / 9.0, 1.0 / 10.0, 1.0 / 3.0, 2.0 / 3.0, 1.0 / 5.0, 2.0 / 5.0, 3.0 / 5.0, 4.0 / 5.0, 1.0 / 6.0, 5.0 / 6.0, 1.0 / 8.0, 3.0 / 8.0, 5.0 / 8.0, 7.0 / 8.0]);
+
+const constants = Object.freeze(["π", "e"]);
+
+const constantValues = Object.freeze([Math.PI, Math.E]);
+
+// communication type
+const AUTOCORRECT_CONTENT = "autocorrectContent";
+
+let insertedText; // Last insert text
+let deletedText; // Last deleted text
+let lastTarget; // Last target
+let lastCaretPosition; // Last caret position
+
+let autocomplete = true;
+let quotes = true;
+let fracts = true;
+
+let autocorrections = {};
+
+let longest = 0;
+
+// Regular expressions
+let symbolpatterns = null;
+// Do not autocorrect for these patterns
+let apatterns = null;
+
+let emojiShortcodes = {};
+
+/**
+ * Get caret position.
+ *
+ * @param {Object} target
+ * @returns {number}
+ */
+function getCaretPosition(target) {
+	if (target.isContentEditable) {
+		target.focus();
+		let _range = document.getSelection().getRangeAt(0);
+		let range = _range.cloneRange();
+		let temp = document.createTextNode("\0");
+		range.insertNode(temp);
+		const caretposition = target.innerText.indexOf("\0");
+		temp.parentNode.removeChild(temp);
+		return caretposition;
+	}
+	else
+		return target.selectionStart;
+}
+
+/**
+ * Insert at caret.
+ * Adapted from: https://www.everythingfrontend.com/posts/insert-text-into-textarea-at-cursor-position.html
+ *
+ * @param {Object} target
+ * @param {string} atext
+ * @returns {void}
+ */
+function insertCaret(target, atext) {
+	const isSuccess = document.execCommand("insertText", false, atext);
+
+	// Firefox input and textarea fields: https://bugzilla.mozilla.org/show_bug.cgi?id=1220696
+	if (!isSuccess && typeof target.setRangeText === "function") {
+		const start = target.selectionStart;
+		target.setRangeText(atext);
+
+		target.selectionStart = target.selectionEnd = start + atext.length;
+
+		// Notify any possible listeners of the change
+		const e = document.createEvent("UIEvent");
+		e.initEvent("input", true, false);
+		target.dispatchEvent(e);
+	}
+}
+
+/**
+ * Count Unicode characters.
+ * Adapted from: https://blog.jonnew.com/posts/poo-dot-length-equals-two
+ *
+ * @param {string} str
+ * @returns {number}
+ */
+function countChars(str) {
+	// removing the joiners
+	const split = str.split("\u{200D}");
+	let count = 0;
+
+	for (const s of split) {
+		// removing the variation selectors
+		count += Array.from(s.split(/[\ufe00-\ufe0f]/).join("")).length;
+	}
+
+	return count;
+}
+
+/**
+ * Delete at caret.
+ *
+ * @param {Object} target
+ * @param {string} atext
+ * @returns {void}
+ */
+function deleteCaret(target, atext) {
+	const count = countChars(atext);
+	if (count > 0) {
+		const isSuccess = document.execCommand("delete", false);
+		if (isSuccess)
+			for (let i = 0; i < count - 1; ++i) {
```

Comment:
Please… this is not Python… :stuck_out_tongue_winking_eye: use brackets here, especially if a multi-line for loop follows (even though syntactically it works yeah, but this increases readability)